pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,638 @@
+
+func (s3 *S3ExportProcess) Start() {
+	s3.startExportProcess()
+}
+
+func (s3 *S3ExportProcess) Stop() {
+	s3.stopExportProcess(true)
+}
+
+func (s3 *S3ExportProcess) startExportProcess() {
+	s3.mutex.Lock()
+	defer s3.mutex.Unlock()
+	if s3.exportProcessRunning {
+		return
+	}
+	s3.exportProcessRunning = true
+	s3.commitTicker = time.NewTicker(s3.commitInterval)
+	s3.uploadCh = make(chan bool, 1)
+	s3.stopCh = make(chan stopPayload, 1)
+	s3.exportWg.Add(1)
+	go func() {
+		defer s3.exportWg.Done()
+		s3.flowRecordPeriodicCommit()
+	}()
+}
+
+func (s3 *S3ExportProcess) stopExportProcess(flushQueue bool) {
+	s3.mutex.Lock()
+	defer s3.mutex.Unlock()
+	if !s3.exportProcessRunning {
+		return
+	}
+	s3.exportProcessRunning = false
+	defer s3.commitTicker.Stop()
+	s3.stopCh <- stopPayload{
+		flushQueue: flushQueue,
+	}
+	s3.exportWg.Wait()
+}
+
+func (s3 *S3ExportProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting S3 exporting process")
+	logTicker := time.NewTicker(time.Minute)
+	defer logTicker.Stop()
+	committedRec := 0
+	ctx := context.Background()
+	for {
+		select {
+		case stop := <-s3.stopCh:
+			klog.InfoS("Stopping S3 exporting process")
+			close(s3.uploadCh)
+			if !stop.flushQueue {
+				return
+			}
+			ctx, cancelFn := context.WithTimeout(ctx, queueFlushTimeout)
+			defer cancelFn()
+			committed, err := s3.batchCommitAll(ctx, s3Uploader)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchCommitAll on stop")
+			} else {
+				committedRec += committed
+				klog.V(4).InfoS("Total number of records committed to S3", "count", committedRec)
+			}
+			return
+		case <-s3.commitTicker.C:
+			committed, err := s3.batchCommitAll(ctx, s3Uploader)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchCommitAll on triggered timer")
+			}
+			if err == nil {
+				committedRec += committed
+			}
+		case <-s3.uploadCh:
+			committed, err := s3.batchCommitAll(ctx, s3Uploader)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchCommitAll on triggered timer")
+			}
+			if err == nil {
+				committedRec += committed
+			}
+			if s3.commitTicker != nil {
+				s3.mutex.Lock()
+				s3.commitTicker.Reset(s3.commitInterval)
+				s3.mutex.Unlock()
+			}
+		case <-logTicker.C:
+			klog.V(4).InfoS("Total number of records committed to S3", "count", committedRec)
+			committedRec = 0
+		}
+	}
+}
+
+// batchCommitAll commits all flow records cached in local deque in one uploadFile
+// call. Returns the number of records successfully committed, and error if
+// encountered. Cached records will be removed only after a successful commit.
+func (s3 *S3ExportProcess) batchCommitAll(ctx context.Context, uploaderAPI S3UploaderAPI) (int, error) {
+	// populate records from deque
+	s3.dequeMutex.Lock()
+	currSize := s3.deque.Len()
+	if currSize == 0 {
+		s3.dequeMutex.Unlock()
+		return 0, nil
+	}
+	// Clear the previous records in expiredRecords slice. Allocated memory is kept.
+	s3.expiredRecords = s3.expiredRecords[:0]
```

Comment:
I don't see the need for another temporary storage here. The records can be taken from the queue and serialized into the buffer directly. There is no need for this slice.

Reply:
Thinking about it more, I don't see the need for a queue either: you can just serialize directly to the buffer in `CacheRecord`. This will reduce overall resource utilization for this process.

Reply:
Thanks. Removed queue. I did one more change which I'm not sure about.

I also removed the `updateCh`, which was previously used to receive the "queue full" signal and trigger an upload. Because if we use buffer as the cache, the whole `batchUploadAll` will need to hold the lock*, we actually can let `CacheRecord` directly call `batchUploadAll` when the buffer size reaches maximum. 

*Just realized to avoid holding the lock during upload, we can create a copy of buffer and upload the buffer, but during the upload, shall we reset the buffer so that it allows new write-in? What if the upload fails, then we need to recover the previous cache data? Would like to hear your opinion.

Reply:
`CacheRecord` should not block, or it impacts other exporters. So `CacheRecord` should not be in charge of the upload, or of any expensive operation.

In case of a failed upload, we cannot keep holding to the data forever. You could have something like this:
```go
PendingBuffers []*bytes.Buffer
CurrentBuffer bytes.Buffer
```
`CurrentBuffer` is updated by `CacheRecord`. When the buffer is full, it is appended to `PendingBuffers`.
The upload function is in charge of uploading everything in `PendingBuffers`, with every buffer as a separate CSV file. When the upload succeeds, the buffer is removed from `PendingBuffers`. `PendingBuffers` should have a max length (let's say 3): after that we drop the oldest buffer (which is the first one in the list).

The logic to "allocate" the new `CurrentBuffer` could look like this:
```go
var releasedBuffer *bytes.Buffer
for len(PendingBuffers) >= 3 {
        releasedBuffer = PendingBuffers[0]
        PendingBuffers = PendingBuffers[1:]
}
newBuffer := releasedBuffer
if newBuffer == nil {
        newBuffer = &bytes.Buffer{}
        // avoid too many memory allocations
        newBuffer.Grow(CurrentBuffer.Cap())
}
PendingBuffers = append(PendingBuffers, CurrentBuffer)
CurrentBuffer = newBuffer
```

Reply:
thanks, changed as suggested